Added Navigators to handle navigation programatically

mcousillas6 opened the pull request:
#### Description:

* Added helpers and basic structure to handle navigation programatically, without the need of using segues.

### Advantages
- No more storyboard segues 🎉 
- One centralized place to initialise your views and viewModels.
- Separation of concerns, the view does not need to know how to init a viewModel, and the viewModel does not need to know about its view.
- Separates the view from the responsability of handling the navigation stack and leaves that to the routes and navigators.
- Reproducible view setups (if you push to the same view from 3 screens, you don't need to repeat the config in the segue).

### Imeplementation details
* The main implementation is in the `Navigator.swift` file, which implements the base behavior for a Navigation.
* Then, The `Route` protocol defines a basic interface to create a UIViewController, your `Routes` are the ones in charge of Creating the UIViewController instance and wire it up with it's corresponding viewModel. Check `OnboardingRoutes` or `HomeRoutes` for examples.
* Semantically, A Navigator is in charge of one NavigationStack. So for example, you will have an `AppNavigator` that manages your root navigation stack for your app and then you could have extra navigator for other flows such as a modal flow that is isolated, or one for each tab on a tabViewController.

germanStabile commented on AcceptanceTests/Tests/SignInTests.swift:
```diff
@@ -30,17 +28,14 @@ class SignInTests: KIFTestCase {
-    if let navigationController = UIApplication.shared.keyWindow?.rootViewController as? UINavigationController {
-      navigationController.popToRootViewController(animated: true)
-    }
+    AppNavigator.shared.navigate(to: OnboardingRoutes.firstScreen, with: .changeRoot, animated: false)
```
should be pop to root

germanStabile commented on ios-base/AppDelegate.swift:
```diff
@@ -30,11 +30,11 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
-      self.window?.rootViewController = vc
-    }
+
+    window = UIWindow(frame: UIScreen.main.bounds)
```
don't think we need to do this

glm4 commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,204 @@
+          self?.rootViewController = navigation
+          self?.currentViewController = self?.rootViewController?.topViewController
+        } else {
+          let navigationController = UINavigationController(rootViewController: viewController)
```
Maybe we can move this few lines into a `embedIntoNavigationController` method since it is repeated.

glm4 commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,204 @@
+      return
+    }
+
+    currentViewController?.present(viewController, animated: animated, completion: completion)
```
I think we can use the method above passing the `TransitionType.modal` to avoid repetitions.

mcousillas6 replied:
Actually, we can't since the navigation method receives a `Route`, here, we are using the rootVc of another navigator.

glm4 replied:
So that leaves routing betweeen `Navigator` objects to be modally only?

germanStabile commented:
> What do you recommend to handle the case when we need to navigate to a certain UIViewController but the view model/params changes?

How about overloading with optional params ?

germanStabile commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,204 @@
+}
+
+public extension Navigator {
+  func navigate(to route: Route, with transition: TransitionType, animated: Bool = true, completion: (() -> Void)? = nil) {
```
how would you handle custom transitions ?

mcousillas6 replied:
We need to add a new transitionType, I was thinking about doing something like `TransitonType.custom` and that one would have a configuration callback that receives the presenting and presented viewControllers as parameters, so you can configure delegates, presentationStyles, etc.

mcousillas6 replied:
Ended up doing something like this, but at `Route` level, I think it makes more sense, since the view is the one who knows the actual views.

germanStabile commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,204 @@
+  func popTo(route: Route, animated: Bool = true) {
+    guard
+      let viewControllers = rootViewController?.viewControllers,
+      let viewController = viewControllers.first(where: {
```
you can loose the `where` here.

mcousillas6 replied:
The compiler fails 🤔 I think the `first { ... }` works only for equatable (not sure though)

germanStabile replied:
That doesn't sound right, it is true for `firstIndex(of:)`. anyways not a big deal

germanStabile commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,204 @@
+  func dismiss(animated: Bool = true, completion: (() -> Void)? = nil) {
+    let presentingViewController = currentViewController?.presentingViewController
+    currentViewController?.dismiss(animated: animated, completion: completion)
+    currentViewController = presentingViewController
```
it seems like if the users directly calls dismiss, push, present the Navigator would loose track of the currentViewController. Maybe we can play with the idea of blocking this APIs, not sure if possible 🤔

mcousillas6 replied:
Yeah, that's a problem, i'm not sure if there's a way of listening to navigation events globally (with the notificationCenter for example)

mcousillas6 replied:
Ended up refactoring the currentViewController variable to be a computed one, since the rootViewController already tracks that from the 'visibleViewController` and `topViewController`

germanStabile commented on ios-base/View/Onboarding/OnboardingRoutes.swift:
```diff
@@ -0,0 +1,45 @@
+        return UIViewController()
+      }
+      signUp.viewModel = SignUpViewModelWithEmail()
+      return signUp
```
this switch has the potential to grow a lot, maybe we can move cases closures to functions to keep it light

germanStabile commented on ios-base/ViewModel/ViewModelState.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+enum ViewModelState: Equatable {
```
Nice!

mcousillas6 commented:
> Looks great!
> What do you recommend to handle the case when we need to navigate to a certain `UIViewController` but the view model/params changes?

To pass parameters, the best approach is to put them as parameters on your Route, if you are using enums like I did you end up with routes like `.login(username: String, password: String)` then, you can use that to initialize your viewModel & views properly.

mcousillas6 commented:
Updated with some improvements: 
- Added a `transitionConfigurator` callback for Routes, so you can set transitionDelegates and so on.
- Changed the way `currentViewController` works, so it keeps in sync with the actual navigation hierarchy.

glm4 commented:
> > Looks great!
> > What do you recommend to handle the case when we need to navigate to a certain `UIViewController` but the view model/params changes?
> 
> To pass parameters, the best approach is to put them as parameters on your Route, if you are using enums like I did you end up with routes like `.login(username: String, password: String)` then, you can use that to initialize your viewModel & views properly.

Do you thing that creating two different routes for the same view but responding to different view models would be a better approach?

mcousillas6 commented:
> > > Looks great!
> > > What do you recommend to handle the case when we need to navigate to a certain `UIViewController` but the view model/params changes?
> > 
> > 
> > To pass parameters, the best approach is to put them as parameters on your Route, if you are using enums like I did you end up with routes like `.login(username: String, password: String)` then, you can use that to initialize your viewModel & views properly.
> 
> Do you thing that creating two different routes for the same view but responding to different view models would be a better approach?

Why would you have 2 view models for the same view? 🤔 Normally, Views and VMs have a 1 to 1 relation, with the VM defining the business behaviour.

Either way, If you really need different VMs, yes, having two separate routes is better. 
I would still go for the parameter approach if you only need to inject some property value, for example, if you have a VM that receives an ID to fetch a profile or similar.

glm4 commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,216 @@
+
+public extension UIViewController {
+  func embedInNavigationController() -> UINavigationController {
+    if let navigation = (self as? UINavigationController) {
```
Please remove enclosing perentheses.

glm4 commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,216 @@
+  func embedInNavigationController() -> UINavigationController {
+    if let navigation = (self as? UINavigationController) {
+      return navigation
+    } else {
```
else not needed here.

glm4 commented:
> > > > Looks great!
> > > > What do you recommend to handle the case when we need to navigate to a certain `UIViewController` but the view model/params changes?
> > > 
> > > 
> > > To pass parameters, the best approach is to put them as parameters on your Route, if you are using enums like I did you end up with routes like `.login(username: String, password: String)` then, you can use that to initialize your viewModel & views properly.
> > 
> > 
> > Do you thing that creating two different routes for the same view but responding to different view models would be a better approach?
> 
> Why would you have 2 view models for the same view? 🤔 Normally, Views and VMs have a 1 to 1 relation, with the VM defining the business behaviour.
> 
> Either way, If you really need different VMs, yes, having two separate routes is better.
> I would still go for the parameter approach if you only need to inject some property value, for example, if you have a VM that receives an ID to fetch a profile or similar.

I meant a view that can be presented in different scenarios using different view models and/or parametes. I think in that case two routes is better to avoid conditionals and extra logic in the same Route case.

mcousillas6 commented:
> > > > > Looks great!
> > > > > What do you recommend to handle the case when we need to navigate to a certain `UIViewController` but the view model/params changes?
> > > > 
> > > > 
> > > > To pass parameters, the best approach is to put them as parameters on your Route, if you are using enums like I did you end up with routes like `.login(username: String, password: String)` then, you can use that to initialize your viewModel & views properly.
> > > 
> > > 
> > > Do you thing that creating two different routes for the same view but responding to different view models would be a better approach?
> > 
> > 
> > Why would you have 2 view models for the same view? 🤔 Normally, Views and VMs have a 1 to 1 relation, with the VM defining the business behaviour.
> > Either way, If you really need different VMs, yes, having two separate routes is better.
> > I would still go for the parameter approach if you only need to inject some property value, for example, if you have a VM that receives an ID to fetch a profile or similar.
> 
> I meant a view that can be presented in different scenarios using different view models and/or parametes. I think in that case two routes is better to avoid conditionals and extra logic in the same Route case.

Oh okay! yes, in that case it's way more clear to have two routes.

germanStabile commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  func popTo(route: Route, animated: Bool = true) {
+    guard
```
I like this guard indentation, @glm4 what do you think about adding this to our style guides ?

germanStabile commented on ios-base/Navigators/Navigator.swift:
```diff
@@ -0,0 +1,215 @@
+ information that it's passed to the Route.
+ */
+public protocol Route {
+  typealias TransitionConfigurator = (_ sourceVc: UIViewController?, _ destinationVc: UIViewController) -> Void
```
🙌

germanStabile commented on ios-base/View/Home/HomeRoutes.swift:
```diff
@@ -0,0 +1,27 @@
+//  ios-base
+//
+//  Created by Mauricio Cousillas on 6/13/19.
+//  Copyright © 2019 TopTier labs. All rights reserved.
```
we need to change the organization name at some point!

mcousillas6 replied:
Maybe in a separate PR, we can change all the files and the config

germanStabile replied:
yes of course, not for this PR

glm4 commented on ios-base/AppDelegate.swift:
```diff
@@ -50,7 +49,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     //Take user to onboarding if needed, do NOT redirect the user if is already in the landing
     // to avoid losing the current VC stack state.
     if window?.rootViewController is HomeViewController {
```
It actually depends on the project but this is done this way because failed login and sign up attepmts are also `unauthorized 401`. Doing so will cause you to be redirected to the landing everytime. For the example app I think is better to check for the Home which indicates that you are logged in, that checking that is not the landing, nor the sign in nor the sign up and each other possible view where you can receive the 401 error and it is in the initial storyboard.

germanStabile replied:
Perhaps we can refactor so sign in/up don't trigger an `unexpectedLogout` but its not the scope of this PR of course

CamilaMoscatelli commented:
Looks great! :clap:
